software/sidekick/sidekick.py

```diff
@@ -0,0 +1,217 @@
+import sys
+import os
+import copy
+import time
+import signal
+import serial
+import serial.tools.list_ports
+import libscrc
+from threading import Thread, Lock, Event
+from enum import Enum
+
+import sign_buddy_pb2 as sbp
+
+sidekick = None
+
+SYNC_BYTE = 0x16
+FRAME_HEADER_SIZE = 6 # SYNC + LENGTH + 4 CRC
+
+class DecoderParseState(Enum):
+    HEADER = 0
+    BODY = 1
+
+# SignBuddy Stream Decoder
+class SBSDecoder():
+    def __init__(self, callback):
+        self.encoded = bytearray()
+        self.state = DecoderParseState.HEADER
+        self.currentFrameSize = 0
+        self.callback = callback
+        self.decodedCRC = 0
+
+    def _decode(self):
+        while True:
+            if self.state == DecoderParseState.HEADER:
+                if len(self.encoded) < FRAME_HEADER_SIZE:
+                    return
+
+                if self.encoded[0] != SYNC_BYTE:
+                    self.encoded = self.encoded[1:]
+                    continue
+
+                self.currentFrameSize = self.encoded[1]
+                self.state = DecoderParseState.BODY
+
+                self.decodedCRC = 0
+                self.decodedCRC |= self.encoded[2] << 0
+                self.decodedCRC |= self.encoded[3] << 8
+                self.decodedCRC |= self.encoded[4] << 16
+                self.decodedCRC |= self.encoded[5] << 24
+
+                # pop out entire header
+                self.encoded = self.encoded[FRAME_HEADER_SIZE:]
+
+            if self.state == DecoderParseState.BODY:
+                if len(self.encoded) < self.currentFrameSize:
+                    return
+
+                frameData = self.encoded[0:self.currentFrameSize]
+
+                if self.decodedCRC != 0xA5A5A5A5: #hard coded value for now
+                    self._soft_reset()
+                    continue
+
+                # shift out data
+                self.encoded = self.encoded[self.currentFrameSize:]
+                # reset for next packet
+                self._soft_reset()
+
+                #pass to decoder
+                self.callback(frameData)
+
+    def _soft_reset(self):
+        self.state = DecoderParseState.HEADER
+        self.currentFrameSize = 0
+        self.decodedCRC = 0
+
+    def decode(self, encoded):
+        self.encoded = self.encoded + encoded
+        self._decode()
+
```

Comment:
I think we should just use one decode function

Reply:
Could move `self.encoded = self.encoded + encoded` inside _decode and rename that to decode